### PR TITLE
Build standalone UPP via cmake

### LINKED ISSUE #246: Update the UPP with libraries from the hpc-stack
Update the UPP modulefiles with libraries from the hpc-stack built on WCOSS Dell, Cray, Hera, Orion and Jet.

### LINKED ISSUE #249: Build UPP standalone executable via cmake
Set up a new function of building UPP standalone executable via cmake instead of GNC makefile.

### PR DESCRIPTION
**Pull request by WenMeng-NOAA**
Add capacity of building standalone UPP via cmake and with all UPP dependency libraries from the hpc-stack on WCOSS-Dell, WCOSS-Cray, Hera, Orion and Jet.   The build script is under tests/complile.sh. The GNC make capacity will be phased out in the future.
The changes include:

- Add new cmake build script at tests/compile_upp.sh
- Add module files  cray-intel, hera, jet, orion and wcoss_dell_p3 from the hpc-stack
- Rename UPP executable to upp.x
- Update version number to 10.0.3

### CODE REVIEW

File: tests/compile_upp.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  exit 2
+else
+  machine=$1 
+fi
+
+case $machine in
+jet)                                   # For Jet
+ module purge
+ . /etc/profile
+ . /etc/profile.d/modules.sh
```
Comment (aerorahul) [inline tests/compile_upp.sh]: why do these files need to be sourced?  Here and on other machines.

File: tests/compile_upp.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  mac=$(hostname | cut -c1-1)
+  mac2=$(hostname | cut -c1-2)
+
+  if [ $mac2 = hf ] ; then   # For Hera
+    machine=hera
+  elif [ $mac = v -o $mac = m  ] ; then  # For WCOSS Dell
+    machine=wcoss_dell_p3
+  elif [ $mac = O ] ; then  # For orion
+    machine=orion
+  elif [ $mac = l -o $mac = s ] ; then  # For WCOSS Cray
```
Comment (aerorahul) [inline tests/compile_upp.sh]: Detecting machine based on first and second character of the hostname is not a good idea.
If building on compute nodes, the hostnames are different.
An alternative is to check for machine specific paths e.g. `/scratch1` or `/gpfs` etc.
See UFS-weather-model for an example of to detect machines and load modules.

File: tests/compile_upp.sh [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+module use ${moduledir}/modulefiles
+module load ${machine}
+module list
+
+set -x
+
+rm -rf build install
+mkdir build && cd build
+cmake -DCMAKE_INSTALL_PREFIX=../install -DBUILD_WITH_WRFIO=ON ../..
+make -j12
```
Comment (aerorahul) [inline tests/compile_upp.sh]: 12 is too many threads to compile in parallel.  I am not sure UPP will benefit significantly from more than 4-6.

File: tests/compile_upp.sh [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ . /etc/profile
+ ;;
+*)
+ set +x
+ echo "ERROR: Invalid machine name specified"
+ usage
+ ;;
+esac
+
+#Load requited modules
```
Comment (aerorahul) [inline tests/compile_upp.sh]: `required` is misspelled.
One should not have to load modules e.g. when in a container or github action

File: tests/compile_upp.sh [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+ . /etc/profile
+ . /etc/profile.d/modules.sh
+ ;;
+orion)                                 # For Orion
+ . /etc/profile
+ ;;
+*)
+ set +x
+ echo "ERROR: Invalid machine name specified"
+ usage
```
Comment (aerorahul) [inline tests/compile_upp.sh]: one should be able to build UPP on any machine, not just the ones listed above

File: tests/compile_upp.sh [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
```
Comment (aerorahul) [inline tests/compile_upp.sh:1]: Why is this script in `tests` directory?  It should be at the top level.

**Review by aerorahul — CHANGES_REQUESTED** [review]
some comments for your consideration.

File: tests/compile_upp.sh (hunk 6 above)
Comment (WenMeng-NOAA) [reply tests/compile_upp.sh:1]: @aerorahul  I use the similar directory structure from the ufs-weather-model. In the future, we would expend more features such as regression test scripts in it.

File: tests/compile_upp.sh (hunk 3 above)
Comment (WenMeng-NOAA) [reply tests/compile_upp.sh]: @aerorahul Would you suggest the option "-j4" or no option? I actually don't know if this option has impact on UPP build performance or run performance.

File: tests/compile_upp.sh (hunk 6 above)
Comment (aerorahul) [reply tests/compile_upp.sh:1]: the ufs `compile.sh` calls top-level `build.sh`

File: tests/compile_upp.sh (hunk 3 above)
Comment (aerorahul) [reply tests/compile_upp.sh]: 4 should be enough. 12 is too much.
`make -jN` enables parallel compilation (as much as the code will allow).

**Comment by WenMeng-NOAA** [comment]
@aerorahul The compile_upp.sh was updated based on your comments. Please review it. Thanks!

**Comment by WenMeng-NOAA** [comment]
The build tests were completed on WCOSS-Dell, WCOSS-Cray, Hera, Orion and Jet. The upp lib and executable were successfully generated.

**Review by aerorahul — APPROVED** [review]
Thanks @WenMeng-NOAA 
This is very much improved. 
In a later PR, I would suggest enabling a choice of a compiler as well as supporting a generic Linux and macOS.

**Comment by WenMeng-NOAA** [comment]
@aerorahul `In a later PR, I would suggest enabling a choice of a compiler as well as supporting a generic Linux and macOS.`
That would be great. Our UPP need more technical supports for improving code performance. Thanks!

**Review by SMoorthi-emc — APPROVED** [review]

**Pull request merged**